### ISSUE TITLE
one-var rule: separate option for function expression assignments?

### ISSUE THREAD

**Issue by aubergine10** [eslint/eslint#6433]
Rule: http://eslint.org/docs/rules/one-var.html

**What version of ESLint are you using?**
2.11.1

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

``` javascript
module.exports = {
  "env": {
    "es6": true,
    "node": true
  },

  "extends": "eslint:recommended",

  "rules": {

    "one-var": [
      "error",
      "always",
    ],

    "space-infix-ops": [
      "error",
      { "int32Hint": false }
    ],

    "template-curly-spacing": [
      "warn",
      "never"
    ],
  }
};
```

**What did you do? Please include the actual source code causing the issue.**

I have a number of `const` definitions, particularly at the top-level of the script but also within functions, many of which are function expression assignments, eg:

``` javascript
const
  foo,
  bar = 1

const baz = function() {
  // ...
}

const quux = (meh) => {
  // ...
}
```

If I use the `uninitialized` and `initialized` options object, I'm forced to either give `bar` it's own `const` declaration, or merge all the function expr assignments in to a single `const` declaration. The same issue arises with the `var`, `let`, `const` options object - there's no way to treat function expr assignments differently.

This is bearable if there are only a few trivial (`bar`-like) assignments, but regularly at the top-level of the script I will have dozens of them. To have my function expr assignments require individual declarations, it means I also have to have dozens of trivial assignments require individual declarations ...that makes the start of scripts look very cluttered due to excess declaration keywords.

There doesn't seem to be a way to set it up so that it's:

> `"always"` (or `"never"`) for everything, _except_ a function expression assignment.

**What did you [want] to happen?**

Ideally I'd like an extra setting in the `var`, `let`, `const` options object: **`functionExpr`**

I could then do something like this:

``` javascript
// ...
  "rules": {

    "one-var": [
      "error",
     {
       "var": "always",
       "let": "always",
       "const": "always",
       "functionExpr": "never", // new option
     },
    ],

// ...
```

The `functionExpr` option could also be applicable to the `uninitialized` and `initialized` options object, but primarily it would be most useful in the `var`, `const`, `let` options object.

A better approach might be to have **`"always-except-fn"`** and **`"never-except-fn"`** values for the `var`, `let` and `const` options? That would allow something like:

``` javascript
    "one-var": [
      "error",
     {
       "var": "always",
       "let": "always",
       "const": "always-except-fn", // new option value
     },
    ],
```

**What actually happened? Please include the actual, raw output from ESLint.**

The rule is working as advertised, for every combination of options, it's just that I want an extra option to handle assignments of function expressions differently without affecting other, more trivial assignments (like the `bar` example earlier).

**Comment by vitorbal** [comment]
Hi @aubergine10, thank you for your proposal! Before we can accept this as an enhancement, we need some 👍s from members of the core team.

I think it would help out if we can fully flesh out how the options "API" would look like. For example I find the following setup a bit ambiguous: `{ "initialized": "always", "functionExpr": "never" }.

**Comment by aubergine10** [comment]
The `"always-except-fn"` and `"never-except-fn"` values could also be used with the `initialized` and `uninitialized` config format... Example of most common use case:

```{ "initialized": "always-except-fn", "uninitialized": "always" }```

**Comment by vitorbal** [comment]
Okay. So `"always-except-fn"` would be an option value that can apply in both option scenarios, right?
I think this extra option makes a lot of sense. @eslint/eslint-team what do you guys think?

@aubergine10 would you be willing to drive this forward and send a pull request for this if it gets accepted?

**Comment by aubergine10** [comment]
Yes, a new value for existing options.

As for a PR, I could try - I've not looked at the code for this specific rule and I'm still very new to eslint so have a bit of a tech deficit to overcome.

**Comment by mikesherov** [comment]
Thanks for contributing! I'm a bit undecided on this option. I have a couple of clarifying questions:

1. Couldn't the same then be said objects, arrays, or really anything multiline? I'm not sure I understand why function expressions get special treatment. Should this rule really say `always-except-multiline`?

2. Are there other ways to handle this: 1. Use function declarations instead (although I suppose you don't get const guarantee), or format your code as such?:

```
const a = 1,
          b = 2,

          c = function() {...
```

3. Is this style popular enough to warrant an option? I don't know that it is or isn't but seems like an edge-case to me.

I can honestly be persuaded here. I'm not asking these questions rhetorically! Thanks in advance!

**Comment by vitorbal** [comment]
I think maybe generalizing to an "always-except-multiline" would give this option more purpose. I can see some value in enforcing a style where multiline expressions always need to have a separate declaration.

I don't particularly follow this kind of style in any of my projects though.

**Comment by nzakas** [comment]
I'm 👎 seems like turning off the rule is a better solution. `one-var` is meant to enforce using just one declaration, and if you want to use multiple, I'm not sure why you'd want to have this rule enabled.

**Comment by aubergine10** [comment]
@mikesherov - yes, "always-except-multiline" makes much more sense.

**Comment by pedrottimark** [comment]
Agree with https://github.com/eslint/eslint/issues/6433#issuecomment-226808127 that the best benefits of this rule are for ES5 code with `var` declarations:

> The single-declaration school of thought is based in pre-ECMAScript 6 behaviors, where there was no such thing as block scope, only function scope. Since all `var` statements are hoisted to the top of the function anyway

To the (admittedly limited) extent that I have studied ES6 code with a maximum of `const` declarations, the configuration seems more like:
one-var: ["error", "never"] // to simplify diffs when declarations are added or deleted
no-use-before-define: "error"

**Comment by aubergine10** [comment]
The more I think about it, it seems what is actually desirable is a `"ignore"` option, so I can do this:

```js
    "one-var": [
      "error",
     {
       "var": "always",
       "let": "always",
       "const": "ignore", // new option value
     },
    ],
```

In the options above, the rule would **not** apply to `const`, but **will** still apply to `var` and `let`.

**Comment by kaicataldo** [comment]
I'm closing this issues because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
